src/policy/v3_policy.cpp

```diff
@@ -0,0 +1,144 @@
+// Copyright (c) 2022 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <policy/v3_policy.h>
+
+#include <coins.h>
+#include <consensus/amount.h>
+#include <logging.h>
+#include <tinyformat.h>
+#include <util/check.h>
+
+#include <algorithm>
+#include <numeric>
+#include <vector>
+
+std::optional<std::tuple<Wtxid, Wtxid, bool>> CheckV3Inheritance(const Package& package)
+{
+    assert(std::all_of(package.cbegin(), package.cend(), [](const auto& tx){return tx != nullptr;}));
+    // If all transactions are V3, we can stop here.
+    if (std::all_of(package.cbegin(), package.cend(), [](const auto& tx){return tx->nVersion == 3;})) {
+        return std::nullopt;
+    }
+    // If all transactions are non-V3, we can stop here.
+    if (std::all_of(package.cbegin(), package.cend(), [](const auto& tx){return tx->nVersion != 3;})) {
+        return std::nullopt;
+    }
+    // Look for a V3 transaction spending a non-V3 or vice versa.
+    std::unordered_map<Txid, Wtxid, SaltedTxidHasher> v3_txid_to_wtxid;
+    std::unordered_map<Txid, Wtxid, SaltedTxidHasher> non_v3_txid_to_wtxid;
+    for (const auto& tx : package) {
+        if (tx->nVersion == 3) {
+            // If duplicate txids exist, this function will still detect violations, but it
+            // will return the earlier transaction's wtxid.
+            Assume(v3_txid_to_wtxid.emplace(tx->GetHash(), tx->GetWitnessHash()).second);
+        } else {
+            Assume(non_v3_txid_to_wtxid.emplace(tx->GetHash(), tx->GetWitnessHash()).second);
+        }
+    }
+    for (const auto& tx : package) {
+        if (tx->nVersion == 3) {
+            for (const auto& input : tx->vin) {
+                if (auto it = non_v3_txid_to_wtxid.find(input.prevout.hash); it != non_v3_txid_to_wtxid.end()) {
+                    return std::make_tuple(it->second, tx->GetWitnessHash(), true);
+                }
+            }
+        } else {
+            for (const auto& input : tx->vin) {
+                if (auto it = v3_txid_to_wtxid.find(input.prevout.hash); it != v3_txid_to_wtxid.end()) {
+                    return std::make_tuple(it->second, tx->GetWitnessHash(), false);
+                }
+            }
+        }
+    }
+    return std::nullopt;
+}
+
+std::optional<std::string> PackageV3SanityChecks(const Package& package)
+{
+    // Check inheritance rules within package.
+    if (const auto inheritance_error{CheckV3Inheritance(package)}) {
```

Comment:
Er, the ugly thing is testmempoolaccept... since `AcceptMultipleTransactions` doesn't actually enforce the child-with-parents thing, we can have this right now:

```python
    @cleanup
    def test_v3_in_testmempoolaccept(self):
        node = self.nodes[0]

        self.log.info("Test that v3 inheritance is accurately assessed in testmempoolaccept")
        tx_v2 = self.wallet.create_self_transfer(version=2)
        tx_v2_from_v2 = self.wallet.create_self_transfer(utxo_to_spend=tx_v2["new_utxo"], version=2)
        tx_v3_from_v2 = self.wallet.create_self_transfer(utxo_to_spend=tx_v2["new_utxo"], version=3)
        tx_v3 = self.wallet.create_self_transfer(version=3)
        tx_v2_from_v3 = self.wallet.create_self_transfer(utxo_to_spend=tx_v3["new_utxo"], version=2)
        tx_v3_from_v3 = self.wallet.create_self_transfer(utxo_to_spend=tx_v3["new_utxo"], version=3)

        # testmempoolaccept paths don't require child-with-parents topology. Ensure that topology
        # assumptions aren't made in inheritance checks.
        test_accept_v2_and_v3 = node.testmempoolaccept([tx_v2["hex"], tx_v3["hex"]])
        assert all([result["allowed"] for result in test_accept_v2_and_v3])

        test_accept_v3_from_v2 = node.testmempoolaccept([tx_v2["hex"], tx_v3_from_v2["hex"]])
        assert all([result["package-error"] == "v3-violation" for result in test_accept_v3_from_v2])

        test_accept_v2_from_v3 = node.testmempoolaccept([tx_v3["hex"], tx_v2_from_v3["hex"]])
        assert all([result["package-error"] == "v3-violation" for result in test_accept_v2_from_v3])

        # Submit the 2 parents, and the children should pass.
        node.sendrawtransaction(tx_v2["hex"])
        node.sendrawtransaction(tx_v3["hex"])
        test_accept_v2_and_v3_children = node.testmempoolaccept([tx_v2_from_v2["hex"], tx_v3_from_v3["hex"]])
        assert all([result["allowed"] for result in test_accept_v2_and_v3_children])
```

But it does blindly check the number of transactions in the package, so maybe you're right and we should just do the simple inheritance check?

Reply:
this looks like a good test to add to tease out the differences in code paths

Reply:
Changed `PackageV3Checks` to properly look at ancestor and descendant sets within package, so it handles packages within testmempoolaccepts now